[analyzer] Modernize, improve and promote chroot checker

vabridgers opened the pull request:
This change modernizes, improves and promotes the chroot checker from alpha to the Unix family of checkers. This checker covers the POS05 recommendations for use of chroot.

The improvements included modeling of a success or failure from chroot and not falsely reporting a warning along an error path. This was made possible through modernizing the checker to be flow sensitive.

vabridgers committed:
[analyzer] Modernize, improve and promote chroot checker

This change modernizes, improves and promotes the chroot checker from
alpha to the Unix family of checkers. This checker covers the POS05
recommendations for use of chroot.

The improvements included modeling of a success or failure from chroot
and not falsely reporting a warning along an error path. This was made
possible through modernizing the checker to be flow sensitive.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/ChrootChecker.cpp:
```diff
@@ -80,19 +88,53 @@ bool ChrootChecker::evalCall(const CallEvent &Call, CheckerContext &C) const {
+          .conjureSymbolVal(nullptr, ChrootCE, LCtx, sIntTy, C.blockCount())
+          .castAs<NonLoc>();
+
+  ProgramStateRef StateChrootFailed, StateChrootSuccess;
+  std::tie(StateChrootFailed, StateChrootSuccess) = state->assume(RetVal);
```
```suggestion
  auto [ChrootFailed, ChrootSucceeded] = state->assume(RetVal);
```

steakhal commented on clang/lib/StaticAnalyzer/Checkers/ChrootChecker.cpp:
```diff
@@ -80,19 +88,53 @@ bool ChrootChecker::evalCall(const CallEvent &Call, CheckerContext &C) const {
+  const TargetInfo &TI = C.getASTContext().getTargetInfo();
+  SValBuilder &SVB = C.getSValBuilder();
+  BasicValueFactory &BVF = SVB.getBasicValueFactory();
+  ConstraintManager &CM = Mgr.getConstraintManager();
```
99% of the cases checkers don't need to interact with the ConstraintManager directly. I'd drop this.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/ChrootChecker.cpp:
```diff
@@ -80,19 +88,53 @@ bool ChrootChecker::evalCall(const CallEvent &Call, CheckerContext &C) const {
+  const llvm::APSInt &Minus1 = BVF.getValue(-1, sIntTy);
+
+  if (StateChrootFailed) {
+    StateChrootFailed = CM.assumeInclusiveRange(StateChrootFailed, RetVal,
+                                                Minus1, Minus1, true);
```
You could just call this on the State. BTW why do we even need this? Could we just call `BindExpr` with `nonloc::ConcreteInt{Minus1}`?

steakhal commented on clang/lib/StaticAnalyzer/Checkers/ChrootChecker.cpp:
```diff
@@ -24,21 +26,30 @@
+// and "jail"
+REGISTER_TRAIT_WITH_PROGRAMSTATE(ChrootState, ChrootKind)
+
+// Track the call expression to chroot for accurate
+// warning messages
+REGISTER_TRAIT_WITH_PROGRAMSTATE(ChrootCall, const Expr *)
```
Why do we need this metadata?
I thought we can already infer this from the State by looking at the exploded node where we changed `ChrootState`, grab the ProgramPoint as a `StmtPoint` and get the Stmt of it. That will be a CallExpr you were looking for.
This should hold because the checker only "acts on" the `eval::Call` and `check::PreCall` events, so transitions can only happen in a `Stmt` ProgramPoint location.

vabridgers replied:
The checker looks for unexpected usages of chdir() that follow an invocation of chroot(), so at the point of detecting an unexpected use of chdir() I want to produce an improved warning that includes a "note" about where chroot() was called. So I stored the Chroot expression with the intent of using it for the warning message. But your comments have got me thinking about simplifying all of this, so will look at doing that. If you have suggestions about simplest and best way, I'd be grateful to hear those. BTW, this pattern also exists in a different checker (cannot remember which one now).

steakhal commented on clang/lib/StaticAnalyzer/Checkers/ChrootChecker.cpp:
```diff
@@ -80,19 +88,53 @@ bool ChrootChecker::evalCall(const CallEvent &Call, CheckerContext &C) const {
+  const QualType sIntTy = C.getASTContext().getIntTypeForBitwidth(
+      /*DestWidth=*/TI.getIntWidth(), /*Signed=*/true);
+
+  const Expr *ChrootCE = Call.getOriginExpr();
+  if (!ChrootCE)
+    return;
+  const auto *CE = cast<CallExpr>(Call.getOriginExpr());
```
We were using CallDescriptions right? Those can only match if the `getOriginExpr()` is non-null CallExpr - if I recall correctly. So this branch should be dead.

vabridgers replied:
See my prior comments. Working on a different approach.

vabridgers replied:
Will address

steakhal commented on clang/lib/StaticAnalyzer/Checkers/ChrootChecker.cpp:
```diff
@@ -80,19 +88,53 @@ bool ChrootChecker::evalCall(const CallEvent &Call, CheckerContext &C) const {
+  }
+
+  if (StateChrootSuccess) {
+    StateChrootSuccess =
+        CM.assumeInclusiveRange(StateChrootSuccess, RetVal, Zero, Zero, true);
```
Same story here. We could just directly call `BindExpr` with `nonloc::ConcreteInt{Zero}`.
After that we wouldn't even need to conjure a new symbol.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/ChrootChecker.cpp:
```diff
@@ -104,15 +146,35 @@ void ChrootChecker::evalChdir(const CallEvent &Call, CheckerContext &C) const {
-      if (Str->getString() == "/")
-        state = Mgr.addGDM(state, ChrootChecker::getTag(),
-                           (void*) JAIL_ENTERED);
+      if (Str->getString() == "/") {
```
I think we could just inline `Str`.
```suggestion
      if (StrRegion->getStringLiteral()->getString() == "/") {
```

vabridgers replied:
Understood, will address.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/ChrootChecker.cpp:
```diff
@@ -121,17 +183,40 @@ void ChrootChecker::checkPreCall(const CallEvent &Call,
-        C.emitReport(
-            std::make_unique<PathSensitiveBugReport>(BT_BreakJail, Msg, N));
-      }
+  const ChrootKind k = C.getState()->get<ChrootState>();
```
The llvm coding style prefers upper-case variable names.

vabridgers replied:
Understood, will address.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/ChrootChecker.cpp:
```diff
@@ -121,17 +183,40 @@ void ChrootChecker::checkPreCall(const CallEvent &Call,
-            std::make_unique<PathSensitiveBugReport>(BT_BreakJail, Msg, N));
-      }
+  const ChrootKind k = C.getState()->get<ChrootState>();
+  if (k == ROOT_CHANGED) {
```
AFAICS we only do anything if this condition is met. I'd prefer an early return instead to reduce the indentation for the rest of the code.

vabridgers replied:
Understood. Will address.

steakhal reviewed (changes requested):
This checker deserved some love for sure. Thank you for pushing for this.
I left quite a few comments, touching style and also direction of this patch.
Thanks Vince!

steakhal commented:
> @steakhal and @NagyDonat , thanks for the comments. I'll address and update the patch. Best

@vabridgers Please avoid force pushed. Prefer `merge` over `rebase` while doing the reviews. It's okay to have a sequence of commits fixing up certain behavior. Once the review is done, the PR is anyways squashed when merging. This helps to crossreference existing comments to your PR, that would GitHub loose track of every time you do a force push.

vabridgers committed:
[analyzer] Modernize, improve and promote chroot checker

Updates #1 for original MR per review comments

https://github.com/llvm/llvm-project/pull/117791#discussion_r1860209751
* Update Release notes with specific gh# addressed, and why these changes
  are made.

https://github.com/llvm/llvm-project/pull/117791#discussion_r1860210481
https://github.com/llvm/llvm-project/pull/117791#discussion_r1860215842
* Update checkers.rst documentation on Chroot checker per review comments.
  Describe what the checker does and reference to POS05-C.

https://github.com/llvm/llvm-project/pull/117791#discussion_r1860233866
* Update use of raw string literals per review comments.

https://github.com/llvm/llvm-project/pull/117791#discussion_r1860245965
* Use simpler form of getting IntTy per review comments.

https://github.com/llvm/llvm-project/pull/117791#discussion_r1860248233
* Simplify invocation of conjureSymbolVal per review comments.

https://github.com/llvm/llvm-project/pull/117791#discussion_r1860250243
* Simplify invocation of state->assume() per review comments.

https://github.com/llvm/llvm-project/pull/117791#discussion_r1860253486
https://github.com/llvm/llvm-project/pull/117791#discussion_r1860256509
https://github.com/llvm/llvm-project/pull/117791#discussion_r1860268803
* Remove use of CM, assumeInclusiveRange(), bind return values
  directly per review comments, simplifying modeling of chroot()
  behavior.

vabridgers commented:
> @vabridgers Please avoid force pushes. 
I understand and will abide by this request. I had tried rebasing and pushing at one time and found I needed to force a push upload my newest changes and thought that was the default flow.

vabridgers commented on clang/lib/StaticAnalyzer/Checkers/ChrootChecker.cpp:
```diff
@@ -104,15 +146,35 @@ void ChrootChecker::evalChdir(const CallEvent &Call, CheckerContext &C) const {
+  // When bug type is resource leak, exploded node N may not have state info
+  // for leaked file descriptor, but predecessor should have it.
+  if (!State->get<ChrootCall>())
+    N = N->getFirstPred();
```
Your comments have inspired different thoughts about how to address the problem I'm wanting to solve. Basically, when the checker detects an unexpected use of chdir() I need information about where chroot() was use prior for a detailed and useful warning message. You probably recognize, but I reused design/use patterns from the stream checker that seemed close to what I want. The solution can evolve of course. Thanks for your comments.

vabridgers replied:
I've created a new approach using a visitor.

vabridgers commented on clang/lib/StaticAnalyzer/Checkers/ChrootChecker.cpp:
```diff
@@ -104,15 +146,35 @@ void ChrootChecker::evalChdir(const CallEvent &Call, CheckerContext &C) const {
+  if (!State->get<ChrootCall>())
+    N = N->getFirstPred();
+
+  const ExplodedNode *Pred = N;
+  while (N) {
+    State = N->getState();
+    if (!State->get<ChrootCall>())
+      return Pred;
+    Pred = N;
+    N = N->getFirstPred();
+  }
```
See my prior comments. Working on a different approach for the problem.

vabridgers committed:
[analyzer] Modernize, improve and promote chroot checker

Update #2 - simple clang-format update I missed

vabridgers commented on clang/lib/StaticAnalyzer/Checkers/ChrootChecker.cpp:
```diff
@@ -121,17 +183,40 @@ void ChrootChecker::checkPreCall(const CallEvent &Call,
+    const ExplodedNode *ChrootCallNode = getAcquisitionSite(Err, C);
+    assert(ChrootCallNode && "Could not find place of stream opening.");
+
+    PathDiagnosticLocation LocUsedForUniqueing;
+    if (const Stmt *ChrootStmt = ChrootCallNode->getStmtForDiagnostics())
+      LocUsedForUniqueing = PathDiagnosticLocation::createBegin(
+          ChrootStmt, C.getSourceManager(),
+          ChrootCallNode->getLocationContext());
```
I took this pattern from the stream checker, I think. I'm wanting to create a more detailed warning message. The checker determines if chdir() was used in an unexpected way and the warning message needs to include information about a prior call to chroot() as a note and warning referring to chdir(). Now that I think about this, maybe this can be simplified. I'll work on it.

vabridgers replied:
Got it now. This comment will be addressed in the next update.

vabridgers committed:
[analyzer] Modernize, improve and promote chroot checker

Updates #3

https://github.com/llvm/llvm-project/pull/117791#discussion_r1860298423
https://github.com/llvm/llvm-project/pull/117791#discussion_r1860294735
* Use early return, avoid indentation of code per comment

https://github.com/llvm/llvm-project/pull/117791#discussion_r1860290631
* Inline Str per review comment

https://github.com/llvm/llvm-project/pull/117791#discussion_r1861148485
* Simplify use of conjured symbol per review comments

https://github.com/llvm/llvm-project/pull/117791#discussion_r1860267300
* Remove unneeded branch per review comments

https://github.com/llvm/llvm-project/pull/117791#discussion_r1860235575
* Remove unnecessary custom uniquing per review comments.

vabridgers commented on clang/lib/StaticAnalyzer/Checkers/ChrootChecker.cpp:
```diff
@@ -98,9 +98,9 @@ void ChrootChecker::evalChroot(const CallEvent &Call, CheckerContext &C) const {
-  NonLoc RetVal =
-          SVB.conjureSymbolVal(/*SymbolTag=*/nullptr, ChrootCE, LCtx, IntTy, C.blockCount())
-          .castAs<NonLoc>();
+  NonLoc RetVal = SVB.conjureSymbolVal(/*SymbolTag=*/nullptr, ChrootCE, LCtx,
+                                       IntTy, C.blockCount())
+                      .castAs<NonLoc>();
```
Will address. Thanks

vabridgers committed:
[analyzer] Modernize, improve and promote chroot checker

Updates #4

https://github.com/llvm/llvm-project/pull/117791#discussion_r1860231190
https://github.com/llvm/llvm-project/pull/117791#discussion_r1860224501
https://github.com/llvm/llvm-project/pull/117791#discussion_r1860232925
* Moved egraph walk to bug report visitor per review comments
* Some minor text updates for comments and descriptions

vabridgers commented:
@steakhal , I believe all comments are now resolved. Thank you for the thoughtful and detailed comments, look forward to concluding this in the best way possible. Thank you!

steakhal committed:
NFC Properly use BugReporterVisitor

steakhal committed:
NFC Simplify ChrootChecker::evalChdir

steakhal committed:
NFC Simplify ChrootChecker::evalChroot

steakhal committed:
NFC Rename variable

steakhal committed:
NFC Cleanup private fields

steakhal committed:
Make evalChroot and evalChdir actually eval the call when it does so

steakhal committed:
NFC Reflow the header comment

steakhal committed:
NFC Hide ChrootKind

steakhal committed:
NFC Reorder stuff

steakhal committed:
Simplify registerChrootChecker and shouldRegisterChrootChecker

steakhal committed:
NFC Remove redundant type

vabridgers commented:
Hi @steakhal , no problem. Thanks for the proactive help!

steakhal committed:
NFC Prefer the member call over the free function

steakhal committed:
NFC Add final

steakhal committed:
NFC Drop a comment

steakhal committed:
Clarify some docs

vabridgers commented:
Thanks @steakhal !
